### PR TITLE
Add Various ColorPicker shapes

### PR DESCRIPTION
**Pull request by gongpha**
Introduce more choices of ColorPicker. The settings are available in property and editor settings.

ภาพ

**Edit : I think this is enough for usability.**

### CODE REVIEW

**Comment by gongpha** [comment]
The thing that makes me surprised is when color is overbright . . .
ภาพ

**Comment by Calinou** [comment]
Do we really need to implement all the ColorPicker shapes out there? I'd wager the most useful one to be the good old HSV wheel.

**Comment by gongpha** [comment]
> 
> 
> Do we really need to implement all the ColorPicker shapes out there? This could end up making the ColorPicker a huge class.
> 
> To me, it seems the most useful/desired ColorPicker shape is the good old HSV wheel. It solves most issues that the old square picker shape has.


I'm gonna re-think and edit this draft to dump out some TODO. And you just posted recently ^^

I know this is enough for editor usability, Not like painting software. The only target that I'm planned is the switcher only.

**Comment by YeldhamDev** [comment]
A spacing at the bottom of the circles to separate it from the current color would make it look better.

**Comment by fire** [comment]
You need to convert to lab colour. If you're around we can chat on discord.

**Comment by gongpha** [comment]
@fire I prefer to use Discord. `gongpha#0238`

File: scene/gui/color_picker.h [hunk 1]
```diff
@@ -68,13 +83,15 @@ class ColorPicker : public BoxContainer {
 	Size2i ms;
 	bool text_is_constructor = false;
 	int presets_per_row = 0;
+	PickerShapeType picker_type;
```
Comment (groud) [inline scene/gui/color_picker.h:86]: We initialize the members in the .h file now. Also, I think the wheel shape should be made the default, as this is likely the most useful shape IMO.

**Review by groud — CHANGES_REQUESTED** [review]
Globally, the code looks ok to me. 
I think the feature is good to have, the ColorPicker node is used outside of the editor itself, so I'd say it make sense to allow for such modification if needed.

**Comment by fire** [comment]
I have updated the PickerShapeType to be SHAPE_VHS_CIRCLE and rebased on master.

Screen Shot 2021-04-06 at 7 30 19 AM

**Comment by groud** [comment]
> I have updated the PickerShapeType to default to SHAPE_VHS_CIRCLE and rebased on master.

I think the SHAPE_HSV_WHEEL would be a better fit. It's more common and easier to use.

Have you seen with OP first? It's not common and practical to force-push changes to someone else's branch.

**Comment by fire** [comment]
I did a rebase and set the Picker to SHAPE_VHS_CIRCLE, but it seems like there's disagreement. So I'll probably go with the consensus.

My personal preference is to also add HSY' because it deals with > 1.0 values better as shown earlier.

image

image

https://wolthera.info/2014/07/hsi-and-hsy-for-kritas-advanced-colour-selector/

**Comment by fire** [comment]
@gongpha Feel free to take control again.

File: editor/plugins/script_text_editor.cpp [hunk 2]
```diff
@@ -1721,6 +1721,9 @@ void ScriptTextEditor::_enable_code_editor() {
 		color_picker->set_raw_mode(true);
 	}
 
+	int picker_shape = EDITOR_GET("interface/inspector/default_color_picker_shape");
```
Comment (groud) [inline editor/plugins/script_text_editor.cpp:1724]: Think should not be configurable, I don't think there's a real use case for it.
You can simply hardcode it, it should be enough.

**Review by groud — CHANGES_REQUESTED** [review]

File: editor/plugins/script_text_editor.cpp (hunk 2 above)
Comment (gongpha) [reply editor/plugins/script_text_editor.cpp:1724]: Is it OK to use `(EnumType)EDITOR_GET("...").operator int()` for casting enum value ?

Comment (groud) [reply editor/plugins/script_text_editor.cpp:1724]: Oh nevermind in fact. I'm dumb. 
Forget about my comment, I thought the picker shape was to configure the shape of the cursor in the wheel/circle/square, not the whole picker mode.

So yeah, the PR looks good to me then.

**Review by groud — APPROVED** [review]
LGTM

**Review by fire — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Comment by pycbouh** [comment]
Note: this PR introduced a regression in the editor (and likely in user projects to a degree) because it creates shaders in the constructor of the ColorPicker control. In the editor this causes slowdowns/hitching when updating the Inspector dock, likely the more color pickers are involved, the more it hitches.

A fix is in works.

**Pull request merged**